tests/bot/cogs/moderation/test_utils.py

```diff
@@ -0,0 +1,345 @@
+import unittest
+from datetime import datetime
+from typing import Union
+from unittest.mock import AsyncMock
+
+from discord import Embed, Forbidden, HTTPException, NotFound
+
+from bot.api import ResponseCodeError
+from bot.cogs.moderation.utils import (
+    has_active_infraction, notify_infraction, notify_pardon, post_infraction, post_user, send_private_embed
+)
+from bot.constants import Colours, Icons
+from tests.helpers import MockBot, MockContext, MockMember, MockUser
+
+RULES_URL = "https://pythondiscord.com/pages/rules"
+APPEAL_EMAIL = "appeals@pythondiscord.com"
+
+INFRACTION_TITLE = f"Please review our rules over at {RULES_URL}"
+INFRACTION_APPEAL_FOOTER = f"To appeal this infraction, send an e-mail to {APPEAL_EMAIL}"
+INFRACTION_AUTHOR_NAME = "Infraction information"
+INFRACTION_COLOR = Colours.soft_red
+
+INFRACTION_DESCRIPTION_TEMPLATE = (
+    "\n**Type:** {type}\n"
+    "**Expires:** {expires}\n"
+    "**Reason:** {reason}\n"
+)
+
+PARDON_COLOR = Colours.soft_green
+
+
+class ModerationUtilsTests(unittest.IsolatedAsyncioTestCase):
+    """Tests Moderation utils."""
+
+    def setUp(self):
+        self.bot = MockBot()
+        self.member = MockMember(id=1234)
+        self.user = MockUser(id=1234)
+        self.ctx = MockContext(bot=self.bot, author=self.member)
+        self.bot.api_client.get = AsyncMock()
+
+    async def test_user_has_active_infraction(self):
+        """Test does `has_active_infraction` return correct value."""
+        test_cases = [
+            {
+                "args": (self.ctx, self.member, "ban"),
+                "get_return_value": [],
+                "expected_output": False
+            },
+            {
+                "args": (self.ctx, self.member, "ban"),
+                "get_return_value": [{
+                    "id": 1,
+                    "inserted_at": "2018-11-22T07:24:06.132307Z",
+                    "expires_at": "5018-11-20T15:52:00Z",
+                    "active": True,
+                    "user": 1234,
+                    "actor": 1234,
+                    "type": "ban",
+                    "reason": "Test",
+                    "hidden": False
+                }],
+                "expected_output": True
+            }
+        ]
+
+        for case in test_cases:
+            args = case["args"]
+            return_value = case["get_return_value"]
+            expected = case["expected_output"]
+
+            with self.subTest(args=args, return_value=return_value, expected=expected):
+                self.bot.api_client.get.return_value = return_value
+
+                result = await has_active_infraction(*args)
+                self.assertEqual(result, expected)
+
+    async def test_notify_infraction(self):
```

Comment:
This test currently depends on the working of two separate units of code: the coroutines `notify_infraction` and `send_private_embed`. In fact, you currently assert whether or not `notify_infraction` works by looking at something that `send_private_embed` produces. That means that we're not testing those units independently from each other and, if the test fails, we don't know which unit caused it to fail.

While it obviously happens that one unit uses another unit in its execution, I think that in this specific case, it's better to isolate the two by patching the `send_private_embed` coroutine withing this test. That allows us to later assert whether or not `send_private_embed` was awaited with the correct arguments using one of the `unittest.AsyncMock` test method.

```suggestion
    @patch("bot.cogs.moderation.utils.send_private_embed")
    async def test_notify_infraction(self, send_private_embed_mock):
```

Another important characteristic of the coroutine we're testing here is that it returns the return value of the `send_private_embed` coroutine it calls. This is an "essential characteristic" for the features that use this utility, so we need to assert that it indeed does so. That way, if someone happens to accidentally remove the `return` statement from the last line of the coroutine while working on it the tests will reflect that mistake by failing.

Now, we don't actually have to invoke the `send_private_embed` coroutine for that, as that would make it depend on that other unit again; instead, we can simply provide a return value for the mock we created by patching `send_private_embed` and asserting that the unit we're actually testing does indeed pass on that return value by returning it as well.